#################################################################################
# The MIT License                                                               #
# Copyright (c) 2017 Fulcrum Genomics LLC                                       #
# Permission is hereby granted, free of charge, to any person obtaining a copy  #
# of this software and associated documentation files (the "Software"), to deal #
# in the Software without restriction, including without limitation the rights  #
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell     #
# copies of the Software, and to permit persons to whom the Software is         #
# furnished to do so, subject to the following conditions:                      #
# The above copyright notice and this permission notice shall be included in    #
# all copies or substantial portions of the Software.                           #
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR    #
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,      #
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE   #
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER        #
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, #
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN     #
# THE SOFTWARE.                                                                 #
#################################################################################

#################################################################################
# R script to generate QC plots CollectStutterMetrics
#
# Two inputs should be given:
#   1. The path prefix of the outputs generated by  CollectStutterMetrics.
#   2. The pdf in which to store the plots.
#################################################################################


options(warn = -1) # Don't emit warnings, only errors
library(ggplot2)
library(scales)  # Require by ggplot
library(reshape2)
library(gridExtra)

# Standard colors
fgblue  = "#2E9DD7"
fggreen = "#155936"
fgred   = "firebrick3"
fpurple = "purple"

args          = commandArgs(trailingOnly=T)
inputPrefix   = args[1]
outputFile    = args[2]

consensusFile = paste(inputPrefix, ".stutter.tab", sep="")
rawFile       = paste(inputPrefix, ".raw.stutter.tab", sep="")
consensusData = read.table(consensusFile, sep="\t", header=F, stringsAsFactors=FALSE)
rawData       = read.table(rawFile, sep="\t", header=F, stringsAsFactors=FALSE)
startCol      = 3

consensusAlleleFile = paste(inputPrefix, ".allele_counts.tab", sep="")
rawAlleleFile       = paste(inputPrefix, ".raw.allele_counts.tab", sep="")
consensusAlleles = read.table(consensusAlleleFile, sep="\t", header=F, stringsAsFactors=FALSE, fill=T)
rawAlleles       = read.table(rawAlleleFile, sep="\t", header=F, stringsAsFactors=FALSE, fill=T)

pdf(outputFile, width=11, height=8.5)

stopifnot(nrow(consensusData) == nrow(rawData))
consensusX = as.numeric(consensusData[1,startCol:ncol(consensusData)])
rawX       = as.numeric(rawData[1,startCol:ncol(rawData)])
stopifnot(all(consensusX == rawX))

allConsensusY = as.numeric(consensusData[nrow(consensusData),startCol:ncol(consensusData)])
allRawY       = as.numeric(consensusData[nrow(consensusData),startCol:ncol(consensusData)])

# Normalize to 1
allConsensusY = allConsensusY / sum(allConsensusY)
allRawY       = allRawY / sum(allRawY)

alleleTable <- function(allAlleles, name, alleleType) {
    data = allAlleles[allAlleles[,1] == name, startCol:ncol(allAlleles)]
    data = data[1,!is.na(data[1,])]  # remove NAs
    data = data[1, data[1,] != ""]  # remove blanks

    to = max(1, ncol(data))

    alleles = data[1,seq(from=min(1, to), to=to, by=3)]
    repeatLengths = data[1,seq(from=min(2, to), to=to, by=3)]
    counts = data[1,seq(from=min(3, to), to=to, by=3)]

    alleles = as.character(alleles)
    repeatLengths = as.numeric(repeatLengths)
    counts = as.numeric(counts)

    if (all(alleles == "Not Accessible")) {
        df = cbind(repeatLengths, counts)
        colnames(df) = c("Repeat #", paste("Counts (", alleleType, ")", sep=""))
    }
    else {
        df = cbind(alleles, repeatLengths, counts)
        colnames(df) = c(paste("Allele (", alleleType, ")", sep=""), "Repeat #", "Counts")
    }
    if (nrow(df) > 1) {
        df = df[order(-repeatLengths), ]
    }

    # TODO: total counts on the bottom row
    return (df)
}

xMin = min(as.numeric(consensusData[1,startCol:ncol(consensusData)]))
xMax = max(as.numeric(consensusData[1,startCol:ncol(consensusData)]))

mytheme <- gridExtra::ttheme_default(
core = list(fg_params=list(cex = 0.5)),
colhead = list(fg_params=list(cex = 0.5)),
rowhead = list(fg_params=list(cex = 0.5)))

# all but the last row ("all_strs"), skip the header
num_strs = nrow(consensusData) - 2
for (i in 2:(nrow(consensusData)-1)) {
    name = consensusData[[i,1]]
    region = consensusData[[i,2]]
    stopifnot(rawData[i,1] == name)
    print(paste("Plotting ", name, " (", (i-1), "/", num_strs, ")", sep=""))

    # Get the allele counts for each repeat length for the raw and conesnsus data
    cAlleles = alleleTable(consensusAlleles, name, "consensus calls")
    rAlleles = alleleTable(rawAlleles, name, "raw reads")

    consensusY = as.numeric(consensusData[i,startCol:ncol(consensusData)])
    rawY       = as.numeric(rawData[i,startCol:ncol(rawData)])

    # Normalize to 1
    consensusY = consensusY / sum(consensusY)
    rawY       = rawY / sum(rawY)

    df = data.frame(
        stutter      = consensusX,
        strRaw       = rawY,
        allRaw       = allRawY,
        strConsensus = consensusY,
        allConsensus = allConsensusY
    )

    df = melt(df, id.vars="stutter", value.name="counts")

    p = ggplot(df, aes(x=stutter, y=counts, fill=variable)) +
        geom_bar(width = 0.9, position=position_dodge(width = 0.9), stat="identity", colour="black") +
        scale_fill_manual(
            labels=c("STR raw reads", "All raw reads", "STR consensus calls", "All consensus calls"),
            values=c(fgblue, fggreen, fgred, fpurple)
        ) +
        labs(x="Stutter distance (log10 scaled)", y="Normalized counts", title=paste(name, " (", region, ")", sep="")) +
        theme(plot.title = element_text(hjust = 0.5), legend.title=element_blank())

    if (nrow(rAlleles) > 0) {
        p = p + annotation_custom(tableGrob(rAlleles, theme=mytheme), xmin=xMin+1, xmax=xMin+1.5, ymin=0.0, ymax=1.0)
    }
    if (nrow(cAlleles) > 0) {
        p = p + annotation_custom(tableGrob(cAlleles, theme=mytheme), xmin=xMax-1.5, xmax=xMax-0.5,  ymin=0.0, ymax=1.0)
    }

    print(p)
}
dev.off()